[12.0][IMP] - Create contract on sale order order confirmation

sbejaoui opened the pull request:
This PR meant to add this functionalities: 
- [x] Sale order lines contain invoicing recurrence details
- [x] Product has a new field is_contract applied only for type service
- [x] Contract products are linked to contract templates
- [x] On Sale Order confirmation, a contract is created for each contract template used on sale order lines
- [x] Contract product are ignored on invoicing process and pass to nothing to invoice
- [x] A not finished contract can be mentioned on sale order line
- [x] A sale order line linked to a contract will update it and don't create a new one if it had the same template

FIXME:
- [ ] remove change made in .travis.yml to get the #207 on build

Commit:
[ADD] product_contract: Create module
* Add contract functionality to `product.templates`
* Add logic to create contracts from `sale.order` that contains contract products.

oca-transbot committed:
OCA Transbot updated translations from Transifex

lasley committed:
[FIX] product_contract: Fix mock usage in tests

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

flotho committed:
[FIX] contract_sale_generation: it doesn't create sales (#141)

* Change the method called in the view
* Complete the create_invoice method
* Bump version + authoring
* Correct bad call of method
  Small Documentation
* Add super call in python test
* FIX bad field names causing bad quantities in sale.order.line

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-travis committed:
[UPD] Update product_contract.pot

sbejaoui committed:
[HACK] Get PR 207, 203 beofre tavis build

this commit must be reverted

sbejaoui committed:
[MIG] - Product Contract Migration to 12.0

sbejaoui committed:
[IMP] - Product with is_contract can be only of type service

sbejaoui committed:
[IMP] - Change dependencies to contract_sale

sbejaoui committed:
[IMP] - Add recurrence fields to product template and sale order line

sbejaoui committed:
[IMP] - Create contract on sale order confirmation

- On Sale Order confirmation, a contract is created for each contract template used on sale order lines
- A not finished contract can be mentioned on sale order line
- A sale order line linked to a contract will update it and don't create a new one if it had the same template

sbidoul commented on product_contract/models/sale_order_line.py:
```diff
@@ -0,0 +1,115 @@
+    date_end = fields.Date(string='Date End', index=True)
+    recurring_next_date = fields.Date(
+        default=fields.Date.today(), copy=False, string='Date of Next Invoice'
+    )
```
What's the use case for having recurring_next_date on the so line?

sbejaoui replied:
It is more user friendly if we can fill all needed information from the sale order.

But thinking about it now, recurring_next_date should have the same value as start_date in prepaid case and start_date + invoicing_interval for postpaid.

sbidoul replied:
Yes, that's my understanding too: this field is not entered by the user but computed on the contract line, so it should not be on the so line.

sbidoul commented on product_contract/models/sale_order.py:
```diff
@@ -0,0 +1,70 @@
+            "contract.action_account_analytic_sale_overdue_all"
+        ).read()[0]
+        action["domain"] = [
+            ("id", "in", self.order_line.mapped('contract_id').ids)
```
I think it's better to search contracts that have a line linked to a sale order line of ours.

env['account.analytic.account'].search([contract_line_ids.recurring_invoice_line_ids.sale_order_line_id.order_id = self.id])

Such a link should be more robust.

sbidoul commented:
@sbejaoui we need have the reference to the so line on the contract line, and then propagating the so line to the invoice line.

sbejaoui committed:
[IMP] - Link contract line to sale order line

sbejaoui committed:
[FIX] - Remove recurring_next_date from sale order line

recurring_next_date should be computed by contract line to get default value

sbejaoui committed:
[IMP] - Contract product are ignored on invoicing process

- Sale order line for contract product pass to nothing to invoice on order confirmation
- Contract Invoices are linked to sale order line

sbejaoui committed:
[FIX] - Default value for date_start applied on product change

sbejaoui committed:
[ADD] - Add readme directory

sbejaoui committed:
[ADD] setup.py

sbejaoui committed:
[IMP] - Onchange contract product on contract contract and contract template

sbejaoui committed:
[FIX] - Change website to OCA repository url and prefix module name with Recurring

sbejaoui committed:
[FIX] - Compute recurring_next_date before create contract line

gva-acsone commented:
Hi @pedrobaeza and @rafaelbn ,

This is the pull request that regroups most of the functionalities that we would like to implement in the 'OCA modules' related to contracts. We have a meeting with our customer end of this week to validate these changes. If you have some time this week, it would be really interesting to have your feedback on these features, as we would like to implement them in the OCA modules in the upcoming week(s).

As a reminder, the major changes are:

- Put the start/end dates and the recurrence mechanism (prepaid/post paid and recurrence interval) on the contract lines instead of the contract form.
- It is now possible to create a contract from a Sales Order, with contract products coming from both the SO and contract template (and not only from the contract template as it was the case before).
- Users can add SO lines that contain contract products in an existing (active) contract.
- The recurrence type (prepaid/postpaid) and the recurrence interval can be filled on the product form of contract products. These informations are copied on the SO line when the product is selected.
- It is now possible to select start/end dates on SO lines. These informations, as well as the recurrence type and the recurrence interval are copied from the SO line on the contract line once the SO is confirmed.
- Invoices for contracts are also available on the related Sales Order form.
- The date of next invoice is correctly updated once the recurrence interval 'Month last day' is selected.
- The invoicing algorithm has been adapted to keep up with all these changes

There are still 2 tasks we are working on:
- invoicing the last month for contract lines that are post-paid (actually not invoiced as the end date is prior to the last invoice date)
- adding the invoices related to contrats on the customers contact form.


Thanks for your feedback,

Guillaume

sbejaoui committed:
[IMP] - Upsell/Downsell contract from sale order

sbejaoui committed:
[ADD] - Add renewal process with termination notice

sbejaoui committed:
[IMP] - compute date end onchange date start for auto-renew sale order lines

rafaelbn commented on product_contract/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+              "ACSONE SA/NV, "
+              "Odoo Community Association (OCA)",
+    'website': 'https://github.com/oca/contract',
+    'depends': ['product', 'contract_sale'],
```
New dependency with contract_sale

rafaelbn commented on product_contract/models/product_template.py:
```diff
@@ -0,0 +1,75 @@
+    @api.constrains('is_contract', 'type')
+    def _check_contract_product_type(self):
+        """
+        Contract product should be service type
```
Review, this is not working  

> Product has a new field is_contract applied only for type service

![2018-11-20_0-11-08](https://user-images.githubusercontent.com/8736623/48740757-07afb400-ec59-11e8-97bf-3c528da868a4.png)

rafaelbn commented:
Hello, 

Thank you for this nice, great and huge work! :smile: 

Some comments from my side.

Let's say firstable that I've never understand why this product_contract should depends on contract templates. For me is an aditional complexity not needed by default. :cry: 

> Put the start/end dates and the recurrence mechanism (prepaid/post paid and recurrence interval) on the contract lines instead of the contract form

Please don't force to every one to define the recurrent mechanism line by line. This should be an option not the rule. This is not a new development, what happen with the customer using this module that only need to define one single recurring mechanism per contract and not by line? I think you should consider both options (maintain the existing + adding the improvements)

![2018-11-20_0-35-23](https://user-images.githubusercontent.com/8736623/48741843-f8326a00-ec5c-11e8-9097-81b573fed64e.png)


> The recurrence type (prepaid/postpaid) and the recurrence interval can be filled on the product form of contract products. These informations are copied on the SO line when the product is selected.

With this you are forcing to make 2 different products for pre-paid and postpaid. This in an error because same product can be selled to diferent customer with diferent invoicing type. Please don't make this filed mandatory.

![2018-11-20_0-24-51](https://user-images.githubusercontent.com/8736623/48742492-6b3ce000-ec5f-11e8-8812-8ff4a3b572c2.png)


> Put the start/end dates and the recurrence mechanism (prepaid/post paid and recurrence interval) on the contract lines instead of the contract form

Again the same. And what happens with the one who was using this option and the new way is not a option because is not needed and is and a extra job? IMHO I think this should be an option and we should work with product_contract without the details of line by line by default.


Take in care about this. If the contract template don't have defined a Journal (which is OK in my side) then when creating the contract from the sale order you should stablish the first sale journal and not leave empty
![2018-11-20_0-29-36](https://user-images.githubusercontent.com/8736623/48742616-fb7b2500-ec5f-11e8-9551-614810e3420c.png)

rafaelbn reviewed (commented):
I beg to consider a more generic and general approach for product_contract and after add more modules like:

- product_contract_invoice_type (allow to set the invoice type to the product for the customer who need this
- product_contract_dates_by_line
- etc....

If not this module is ALL-IN (all or nothing)

Thank you and sorry for the noise.

I will test more deeper as I've found use cases to consider but next time I will record a video

rafaelbn commented:
Please @sbejaoui could you give me the sector and product examples in your side of your customer? Just to put my self in you situation. I my side we have generally Service Consulting Companies, Software companies that sell product and services. Maintenance companies...

sbejaoui commented:
Hello @rafaelbn ,
thank you for your review, I will take your remark one by one.
To understand our use case, this is an example of contract that need recurrence mechanism on lines level:

- [01/01 - 12/31] - An annual subscription plan for website hosting
- [06/01 - 07/31] - An advertising campaign
- [01/01 - 31/03] - A discount of 30$ 
- [01/09 - 30/09] - A mass mailing campaign

I join @pedrobaeza in his idea, when he says, we can leave the recurring fields on both sides, contract and lines.
Visible by default at the contract level and used to fill lines level.

rafaelbn commented:
OK, thank you @sbejaoui . But again remember (1) you are not creating something from scratch you are improving something that already exists and is in use in production (2) if you don't get a more modular and generic approach you will have another customer with different requirement that this modules will not fit. Just trying to do something generic.

> I join @pedrobaeza in his idea, when he says, we can leave the recurring fields on both sides, contract and lines. Visible by default at the contract level and used to fill lines level.

OK, I agree

Thank you

sbejaoui committed:
[IMP] - link contract line and its successor in upsel case

sbejaoui committed:
[FIX] - on upsel, contract line should stop day - 1

rafaelbn commented:
Hello @sbejaoui 



- [ ] When a Product Type in Consumable the Check box "Is a contract" is available. Should be fixed? :thinking: 
- [ ] As I said we have here the same problem if we want to define one contract template with
  - Invoicing type
  - Repeat every
  - Etc...
We cannot because we have to define all that data product by product

Have do you planned all this staff in the migration from v10? Deleting all the fileds in the contract template and adding new fields in the product

Sorry, really this big changes are being hard, imagine for the customers. I wish we sill not have problems with them.

- [ ] Again if you don't delete the fields from the contract and the contract template and you don't make this new fields mandatory in the product than you can cover you uses cases and customer that don't need recurring by line maintain its functionality from v8/9/10/11

> I join @pedrobaeza in his idea, when he says, we can leave the recurring fields on both sides, contract and lines. Visible by default at the contract level and used to fill lines level.

This is important

Thank you!

sbejaoui commented:
@rafaelbn ,

> * [ ]  When a Product Type in Consumable the Check box "Is a contract" is available. Should be fixed? thinking

Yes, we can do that
> * [ ]  As I said we have here the same problem if we want to define one contract template with
>   
>   * Invoicing type
>   * Repeat every
>   * Etc...
>     We cannot because we have to define all that data product by product

Fields on contract template are default value used to automatically fill on product change. A kind of helper.
You are right, this filelds should not be required.
 
> Have do you planned all this staff in the migration from v10? Deleting all the fileds in the contract template and adding new fields in the product

Fields still exist on contract template, but at lines level and we don't need to bring data to product template.
As discussed, at the end of this work, we will bring recurring fields back to contract and contract template level, and we should have the same functional flows as the actual version but with the possibility to chose to manage contracts globally or line by line.

> Sorry, really this big changes are being hard, imagine for the customers. I wish we sill not have problems with them.

I understand your concern, and we will work it together to have a more generic module that fit customers needs. 

> * [ ]  Again if you don't delete the fields from the contract and the contract template and you don't make this new fields mandatory in the product than you can cover you uses cases and customer that don't need recurring by line maintain its functionality from v8/9/10/11

I agree,

Thank you!

sbejaoui committed:
[FIX] - fix onchange

sbejaoui committed:
[IMP] - hide recurring_invoicing_type if recurring_rule_type is monthlylastday

for the monthlylastday case, pre-paid is logicly impossible,
if monthlylastday is set, we consider only post-paid case

rafaelbn commented:
Thank you to you! :heart:

sbejaoui committed:
[FIX] - include date_end in the period if the product is_autorenew

sbejaoui committed:
[IMP] - _prepare_contract_value for sale order confirm

sbejaoui committed:
[REM] - Remove unused method

sbejaoui committed:
[IMP] - get contract user from sale order user

sbejaoui committed:
[IMP] - show related sale orders in contract form

ThomasBinsfeld committed:
[REF] Contract Product: invoice in prepare_invoice_line is optional

sbejaoui committed:
[FIX] - fix flake8

ThomasBinsfeld committed:
[ADD] Product Contract: payment term

sbejaoui committed:
[IMP] - update invoice_line vals only if it is not null

sbejaoui committed:
[FIX] - fix unit tests

ThomasBinsfeld committed:
[ADD] Contract Sale Payment Mode

sbejaoui committed:
[IMP] - Add unit test

sbejaoui committed:
[REF] - predecessor_contract_line set in create process

sbejaoui committed:
[IMP] - Simplify sale order line creation for contract product

ThomasBinsfeld committed:
[ADD] Contract Sale: fiscal position

sbejaoui committed:
[IMP] - auto_renew_rule deducted from recurring_rule_type and product_uom_qty

ThomasBinsfeld committed:
[FIX] Product Contract: copy invoicing partner to created contract

sbejaoui committed:
[FIX] - Fix recurring_next_date

sbejaoui committed:
[FIX] - date_end required only for contract product

sbejaoui committed:
[FIX] - Get auto_renew_rule_type based on recurring_rule_type

sbejaoui committed:
[IMP] - cancel contract line if the downsell/upsell comes before

sbejaoui committed:
[FIX] - date_end computed only for contract products

ThomasBinsfeld committed:
[REF] Contract: set order count to 0 if no access to orders

sbejaoui committed:
[IMP] - permit contract selection without contract_template_id

AaronHForgeFlow reviewed (commented):
It works good, great feature!

Just a minor comment. If the product in the sale order line is a contract. Then the contract_id field should be readonly to prevent users to fill it. contract_id will be filled on SO confirmation so it is fine.

sbejaoui commented:
> It works good, great feature!
> 
> Just a minor comment. If the product in the sale order line is a contract. Then the contract_id field should be readonly to prevent users to fill it. contract_id will be filled on SO confirmation so it is fine.

User can set contract_id if he want to add a sale line to an existing contract. Contract are filtered by customer and contract template.

A user can also set a contract line in that contract so the system will stop the target at the start of the new one (cases of upsell/downsell)

sbejaoui commented:
In version 11.0, product_contract only adds the creation of a contract from a sales order based on a contract template linked to a product without considering the sale order line it self.

With this PR we are improving and adding many features to the sales order/contracts relationship, such as upsell and downsell, converting orders into contracts and handling invoiced quantity for converted lines.

In the other hand, contract_sale module, add the contract menu to the sales application. 

When I was working on this PR, I always wondered why we keep two separate modules with the same purpose (introduce contract management into the sales application).

So isn't the time now to merge them in one module (contract_sale), or can you see a reason why we should keep them separate?

AaronHForgeFlow commented:
> User can set contract_id if he want to add a sale line to an existing contract. Contract are filtered by customer and contract template.
> 
> A user can also set a contract line in that contract so the system will stop the target at the start of the new one (cases of upsell/downsell)

Yes I agree. Please dismiss my comment

hparfr commented:
I did some functional tests, it looks goods.

Filling date start and date end are painful to enter for users and error prone if they are the same for all the lines.

Last remark, is it possible to rename ['sale.order.line'].date_start  and ['sale.order.line'].date_end to something more obvious like contract_date_start and contract_date_end ?

qgroulard committed:
[ADD] Add start_date to contract_line display_name

In order to help distinguishing lines with the same name

rafaelbn commented:
i set WIP because depends on https://github.com/OCA/contract/pull/207 and bots are :red_circle:

sbejaoui committed:
[12.0][IMP] - option to decide if we automaticallt create contract a sale confirm

sbejaoui committed:
[FIX] - fix display name compute dependecy

sbejaoui committed:
[12.0][FIX] - create contract only on sale and done states

sbejaoui committed:
[12.0][REF] - Split analytic accouts and contracts

sbejaoui commented:
Hello @rafaelbn ,
I made the rebase. This PR is ready to be reviewed.

sbejaoui committed:
[12.0][IMP] - Improve unit tests

sbejaoui committed:
[IMP] - improve is contract visibility in product view

ThomasBinsfeld commented:
Conflicts resolved (on .travis.yml)

sbejaoui committed:
[12.0][IMP] - sale line linked to contracts must not be invoiced from sale order

gva-acsone reviewed (approved):
Functional test + 1

hparfr commented:
A little late on this point, but @sbejaoui, have you considered to depend of this module https://github.com/OCA/sale-workflow/blob/12.0/sale_start_end_dates ?

sbejaoui commented:
> A little late on this point, but @sbejaoui, have you considered to depend of this module https://github.com/OCA/sale-workflow/blob/12.0/sale_start_end_dates ?

Hi @hparfr ,

The same question was raised by @sbidoul [here](https://github.com/OCA/sale-workflow/pull/943#issuecomment-539883414).

I think we should do it. 

This improvement worth a git issue: #395 395.

hparfr commented:
Hi. My bad,  I didn't see it :)
Thanks
